### DISCUSSION TITLE
Field option to encrypt in DB

### DISCUSSION THREAD

**Discussion by TomDo1234** [payloadcms/payload#1644] [Feature Requests & Ideas]
I am planning of making a PR where PayloadCMS fields are given an "encrypted" option to allow them to be encrypted in the DB so if attackers get the DB they still gotta decrypt it.

The encrypt option won't just be true false, it will have additional options such as what key it should encrypt with if you want it to be different to your payload secret.

Is this a good idea? Will only work on it if you guys think it is a good idea.

**Comment by JarrodMFlesch** [comment]
This is interesting, but might get a little tricky with querying data. You could encrypt the data going in with field `afterChange` hook pretty easily - even just using the built in `payload.encrypt`.

I think the tricky part would be querying data, let's say I encrypt a `slug` field. How would I query something like `?where[slug][equals]=home`? I would have to encrypt the slug _before_ querying the database so it knows what its looking for.

We have `beforeOperation` hooks that are available on the collection level, but not the field level (perf reasons). You _could_ use the `beforeOperation` to do this, but you would need to search the query params and then encrypt the key and merge it back into the where query - not impossible, also kinda clumsy feeling and might be better off as something that you build and share but not as a part of core functionality.

Thoughts?

**Reply by jmikrut** [reply]
I have done this before in a project. The trick is to create a "blind index" on the field. It's pretty complex stuff but I've actually done it before. @TomDo1234 you can look at how we store API keys in the database now, because that uses a blind index to encrypt the field. It could be a good PoC for you to look at. It's all done with hooks! 👍
